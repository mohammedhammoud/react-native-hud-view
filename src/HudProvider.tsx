import {
  Animated,
  Easing,
  StyleSheet,
  View,
  ViewProps,
  ViewStyle,
} from 'react-native';
import React, {
  useCallback,
  useContext,
  useMemo,
  useRef,
  useState,
} from 'react';

const styles = StyleSheet.create({
  container: {
    alignItems: 'center',
    bottom: 0,
    flex: 1,
    justifyContent: 'center',
    left: 0,
    position: 'absolute',
    right: 0,
    top: 0,
  },
  iconContainer: {
    alignItems: 'center',
    justifyContent: 'center',
  },
  view: {
    flex: 1,
  },
});

export type IconProps = { color?: unknown; name: string; size?: number };

export type HudOptions = {
  backgroundColor?: string;
  borderRadius?: number;
  fadeDuration?: number;
  height?: number;
  rotate?: boolean;
  rotateDuration?: number;
  width?: number;
};

type HudContextProps<P extends IconProps> = {
  hide: () => void;
  show: (iconProps: P, options?: HudOptions) => void;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const HudContext = React.createContext<HudContextProps<any>>({
  hide: () => undefined,
  show: () => undefined,
});

export const useHudContext = <P extends IconProps>() => {
  const context = useContext<HudContextProps<P>>(HudContext);
  return context;
};

export type HudProviderProps<P extends IconProps> = React.PropsWithChildren & {
  IconComponent: React.ComponentType<P>;
  containerStyles?: ViewStyle;
  useNativeDriver?: boolean;
} & ViewProps &
  HudOptions;

export const HudProvider = <P extends IconProps>({
  IconComponent,
  backgroundColor = 'rgba(0,0,0,0.8)',
  borderRadius = 5,
  children,
  containerStyles,
  fadeDuration = 700,
  height = 80,
  rotate = false,
  rotateDuration = 800,
  style,
  useNativeDriver = true,
  width = 80,
  ...viewProps
}: HudProviderProps<P>) => {
  const fadeAnimationRef = useRef(new Animated.Value(0));
  const rotationAnimationRef = useRef(new Animated.Value(0));
  const [iconProps, setIconProps] = useState<P>();

  const defaultOptions = useMemo<Required<HudOptions>>(
    () => ({
      backgroundColor,
      borderRadius,
      fadeDuration,
      height,
      rotate,
      rotateDuration,
      width,
    }),
    [
      fadeDuration,
      backgroundColor,
      borderRadius,
      height,
      width,
      rotate,
      rotateDuration,
    ]
  );

  const [options, setOptions] = useState<HudOptions>(defaultOptions);

  const fadeIn = useCallback(() => {
    Animated.timing(fadeAnimationRef.current, {
      duration: options.fadeDuration,
      toValue: 1,
      useNativeDriver,
    }).start();
  }, [options.fadeDuration, useNativeDriver]);

  const fadeOut = useCallback(() => {
    Animated.timing(fadeAnimationRef.current, {
      duration: options.fadeDuration,
      toValue: 0,
      useNativeDriver,
    }).start(({ finished }) => {
      if (finished) {
        rotationAnimationRef.current.stopAnimation();
        setIconProps(undefined);
      }
    });
  }, [options.fadeDuration, useNativeDriver]);

  const startRotating = useCallback(() => {
    rotationAnimationRef.current.setValue(0);

    Animated.loop(
      Animated.timing(rotationAnimationRef.current, {
        duration: options.rotateDuration,
        easing: Easing.linear,
        toValue: 1,
        useNativeDriver,
      })
    ).start();
  }, [options.rotateDuration, useNativeDriver]);

  const show = useCallback<HudContextProps<P>['show']>(
    (nextIconProps, incomingOptions) => {
      fadeIn();

      const filteredIncomingOptions = incomingOptions
        ? Object.fromEntries(
            Object.entries(incomingOptions).filter(
              ([, value]) => value !== undefined
            )
          )
        : undefined;

      const nextOptions = { ...defaultOptions, ...filteredIncomingOptions };
      setOptions(nextOptions);

      if (nextOptions.rotate) {
        startRotating();
      }

      setIconProps({
        color: 'white',
        size: Math.min(nextOptions.width, nextOptions.height) * 0.5,
        ...nextIconProps,
      });
    },
    [defaultOptions, fadeIn, startRotating]
  );

  const hide = useCallback(() => {
    fadeOut();
  }, [fadeOut]);

  const iconContainerStyles = useMemo(
    () => [
      styles.iconContainer,
      {
        backgroundColor: options.backgroundColor,
        borderRadius: options.borderRadius,
        height: options.height,
        opacity: fadeAnimationRef.current,
        width: options.width,
      },
      containerStyles,
    ],
    [
      containerStyles,
      options.backgroundColor,
      options.borderRadius,
      options.height,
      options.width,
    ]
  );

  const iconWrapperStyles = useMemo(
    () =>
      options?.rotate
        ? {
            transform: [
              {
                rotate: rotationAnimationRef.current.interpolate({
                  inputRange: [0, 1],
                  outputRange: ['0deg', '360deg'],
                }),
              },
            ],
          }
        : {},
    [options?.rotate]
  );

  return (
    <HudContext.Provider value={{ hide, show }}>
      <View
        style={StyleSheet.flatten([styles.view, style])}
        testID="hud-view"
        {...viewProps}
      >
        {children}
        {iconProps ? (
          <View
            pointerEvents="none"
            style={styles.container}
            testID="hud-container"
          >
            <Animated.View
              style={iconContainerStyles}
              testID="hud-icon-container"
            >
              <Animated.View
                style={iconWrapperStyles}
                testID="hud-icon-wrapper"
              >
                <IconComponent {...iconProps} />
              </Animated.View>
            </Animated.View>
          </View>
        ) : null}
      </View>
    </HudContext.Provider>
  );
};
