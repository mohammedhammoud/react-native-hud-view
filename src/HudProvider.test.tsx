import { HudProvider, useHudContext } from './HudProvider';
import { TouchableOpacity, View } from 'react-native';
import { act, fireEvent, render } from '@testing-library/react-native';

import React from 'react';
import renderer from 'react-test-renderer';

const MockIcon = () => <></>;

const Content = () => {
  const { hide, show } = useHudContext();

  return (
    <View testID="content">
      <TouchableOpacity
        onPress={() => show({ name: 'test-icon' })}
        testID="show-hud-button"
      />
      <TouchableOpacity onPress={() => hide()} testID="hide-hud-button" />
    </View>
  );
};

describe('HudProvider', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  it('renders correctly', () => {
    const json = renderer
      .create(
        <HudProvider IconComponent={MockIcon}>
          <View testID="content" />
        </HudProvider>
      )
      .toJSON();
    expect(json).toMatchSnapshot();
  });

  it('renders correctly when toggling hud', () => {
    const { getByTestId, toJSON } = render(
      <HudProvider
        IconComponent={MockIcon}
        fadeDuration={1000}
        testID="hud-provider"
      >
        <Content />
      </HudProvider>
    );

    const showHudButton = getByTestId('show-hud-button');
    const hideHudButton = getByTestId('hide-hud-button');

    fireEvent.press(showHudButton);
    expect(toJSON()).toMatchSnapshot();

    fireEvent.press(hideHudButton);

    act(() => {
      jest.advanceTimersByTime(1000);
    });

    expect(toJSON()).toMatchSnapshot();
  });

  it('should render children', () => {
    const { getByTestId } = render(
      <HudProvider IconComponent={MockIcon}>
        <View testID="content" />
      </HudProvider>
    );
    expect(getByTestId('content')).toBeDefined();
  });

  it('should show and hide hud', () => {
    const { getByTestId } = render(
      <HudProvider
        IconComponent={MockIcon}
        fadeDuration={1000}
        testID="hud-provider"
      >
        <Content />
      </HudProvider>
    );

    const hudProvider = getByTestId('hud-provider');
    const showHudButton = getByTestId('show-hud-button');
    const hideHudButton = getByTestId('hide-hud-button');

    expect(hudProvider.children).toHaveLength(1);

    fireEvent.press(showHudButton);

    expect(hudProvider.children).toHaveLength(2);

    fireEvent.press(hideHudButton);

    jest.advanceTimersByTime(999);
    expect(hudProvider.children).toHaveLength(2);

    act(() => {
      jest.advanceTimersByTime(1000);
    });

    expect(hudProvider.children).toHaveLength(1);
  });

  it('should render hud with custom properties', async () => {
    const IconComponent = jest.fn((props) => <View {...props} />);
    const { getByTestId } = render(
      <HudProvider
        IconComponent={IconComponent}
        backgroundColor="red"
        borderRadius={10}
        height={50}
        rotate={true}
        width={100}
      >
        <Content />
      </HudProvider>
    );

    const showButton = getByTestId('show-hud-button');

    fireEvent.press(showButton);

    expect(IconComponent).toHaveBeenCalledWith(
      {
        color: 'white',
        name: 'test-icon',
        size: 25,
      },
      {}
    );

    expect(getByTestId('hud-icon-container')).toHaveStyle({
      backgroundColor: 'red',
      borderRadius: 10,
      height: 50,
      width: 100,
    });

    expect(getByTestId('hud-icon-wrapper')).toHaveStyle({
      transform: [{ rotate: '0deg' }],
    });
  });
});
