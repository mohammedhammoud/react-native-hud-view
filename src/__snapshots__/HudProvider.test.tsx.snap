// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HudProvider renders correctly 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
  testID="hud-view"
>
  <View
    testID="content"
  />
</View>
`;

exports[`HudProvider renders correctly when toggling hud 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
  testID="hud-provider"
>
  <View
    testID="content"
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
      testID="show-hud-button"
    />
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
      testID="hide-hud-button"
    />
  </View>
  <View
    pointerEvents="none"
    style={
      {
        "alignItems": "center",
        "bottom": 0,
        "flex": 1,
        "justifyContent": "center",
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
    testID="hud-container"
  >
    <View
      collapsable={false}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "rgba(0,0,0,0.8)",
          "borderRadius": 5,
          "height": 80,
          "justifyContent": "center",
          "opacity": 0,
          "width": 80,
        }
      }
      testID="hud-icon-container"
    >
      <View
        collapsable={false}
        style={{}}
        testID="hud-icon-wrapper"
      />
    </View>
  </View>
</View>
`;

exports[`HudProvider renders correctly when toggling hud 2`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
  testID="hud-provider"
>
  <View
    testID="content"
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
      testID="show-hud-button"
    />
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
      testID="hide-hud-button"
    />
  </View>
</View>
`;
